modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatGeobufFeatureReader.java

```diff
@@ -82,28 +86,55 @@ public class FlatGeobufFeatureReader implements FeatureReader<SimpleFeatureType,
             skipNBytes(inputStream, headerMeta.offset);
         }
 
+        Filter filter = q.getFilter();
         Envelope bbox = new ReferencedEnvelope();
-        if (q != null && q.getFilter() != null) {
-            bbox = (Envelope) q.getFilter().accept(ExtractBoundsFilterVisitor.BOUNDS_VISITOR, bbox);
+        Id id = null;
+        if (q != null && filter != null) {
+            bbox = (Envelope) filter.accept(ExtractBoundsFilterVisitor.BOUNDS_VISITOR, bbox);
+            if (filter instanceof Id) id = (Id) filter;
         }
         if (bbox == null
                 || bbox.isNull()
                 || Double.isInfinite(bbox.getWidth())
                 || Double.isInfinite(bbox.getHeight())) {
             bbox = null;
         }
+        if (bbox != null)
+            it =
+                    FeatureCollectionConversions.deserialize(
+                                    inputStream, headerMeta, featureType, bbox)
+                            .iterator();
+        else if (id != null) {
+            long[] fids = extractSortedFids(id);
+            it =
+                    FeatureCollectionConversions.deserialize(
+                                    inputStream, headerMeta, featureType, fids)
+                            .iterator();
+        } else {
+            it =
+                    FeatureCollectionConversions.deserialize(inputStream, headerMeta, featureType)
+                            .iterator();
+        }
+    }
+
+    private static long[] extractSortedFids(Id id) {
+        long[] fids = id.getIdentifiers().stream().mapToLong(i -> extractFid(i)).toArray();
+        Arrays.sort(fids);
+        return fids;
+    }
 
-        it =
-                FeatureCollectionConversions.deserialize(inputStream, headerMeta, featureType, bbox)
-                        .iterator();
+    private static long extractFid(Identifier i) {
```

Comment:
Native feature id in FlatGeobuf are longs so I have to extract it from Identifer and I found no other way to get rid of the namespace prefix. Is this ok @aaime or do you know of a better way?

Reply:
Good points. Added tests for unexpected id and out of bounds id. I think the exceptions thrown make sense.

Reply:
Reworked to filter out instead of throw.